osu.Game/Rulesets/UI/RulesetInputManager.cs

```diff
@@ -166,7 +166,7 @@ public void Attach(KeyCounterDisplay keyCounter)
                                                    .Select(b => b.GetAction<T>())
                                                    .Distinct()
                                                    .OrderBy(action => action)
-                                                   .Select(action => new KeyCounterAction<T>(action)));
+                                                   .Select(action => keyCounter.CreateKeyCounter(new KeyCounterActionTrigger<T>(action))));
```

Comment:
Sorry but I'm just not feeling this. It all seems back to front.

Whenever you see a class passing something to itself *created by a method on that class* you know something's up.

This needs to either:

- Change back to `keyCounter.Add(new KeyCounterAction<T>` (what it was before), moving actual implementation differences out of that class and making it a model class (ie. non-drawable).
- Change to `keyCounter.AddCounter<T>(action)`.

@bdach could you please confirm you agree that the way this PR is currently heading feels wrong?

Reply:
Agree that this is too weird to live.

Either resolution sounds fine, but I will note that in general I'm really not sure how long `KeyCounterDisplay` will live in the state proposed in this PR wherein it's just a fill flow with separate `KeyCounter`s. This is going to make a lot of stuff difficult to do, including for instance the legacy key counter. To me the key counter display (as a whole, all of its pieces) should be a single interchangeable skinnable and the simple flow layout should be moved out to implementations. Which I guess is probably closer to option 1 in general direction.

Reply:
So this means that there will be a single `KeyCounterDisplay` and we're moving back to the old `KeyCounter` inheritance scheme, is that right? 

If that's the case, then how to provide the individual _display implementations_?

I'm not quite sure if I fully understood your directions...

Reply:
Very fast and loose sketch of how I imagine this working:

```csharp
public abstract class KeyCounterDisplay : CompositeDrawable
{
	public readonly Bindable<bool> AlwaysVisible { get; }
	public readonly bool IsCounting { get; set; } // maybe consider bindable after all?
	
	protected readonly Bindable<bool> ConfigVisibility;
	
	protected abstract void UpdateVisibility();
	
	public void SetReceptor(Receptor receptor);
	public void AddTrigger(KeyCounter.InputTrigger trigger); // probably move out InputTrigger so that it's not an inner class anymore
	
	protected abstract KeyCounter CreateKeyCounter(KeyCounter.InputTrigger trigger);
	
	// notably, Content, KeyFlow, etc. are not here - this class has no layout logic, it just hooks up data flows
}

public class ArgonKeyCounterDisplay : KeyCounterDisplay { // ... }
public class TrianglesKeyCounterDisplay : KeyCounterDisplay { // ... }
public class LegacyKeyCounterDisplay : KeyCounterDisplay { // ... }
```

Would like @peppy's opinion before you act on it though.

Reply:
I thought at some point of doing something like this, but I guess I overengineered things in my head which led me to put this idea aside.

Reply:
This might be okay, so long as `new InputTrigger()` is possible.

Reply:
Do you mean that I should make `InputTrigger` not abstract with a public parameterless constructor?

Reply:
It can be abstract as long as it's not per-skin.

Reply:
Oh okay. Well, it has never been the goal to make it skin dependent and will never be, they are just only used to trigger Activation and Deactivation states for the `KeyCounter` to react to. 
Also see the recent changes, I've applied the suggestions from bdach.